File: queue.py
# A queue is a useful data structure in programming. 
# It is similar to the ticket queue outside a cinema hall, where the first person entering 
# the queue is the first person who gets the ticket.
# Queue follows the First In First Out (FIFO) rule 
# - the item that goes in first is the item that comes out first.
# In programming terms, putting items in the queue is called enqueue, 
# and removing items from the queue is called dequeue.

# Basic operations that are allowed on queue
# A Queue is an object (an abstract data structure - ADT) that allows the following operations.

# Enqueue: Add an element to the end of the queue
# Dequeue: Remove an element from the front of the queue
# IsEmpty: Check if the queue is empty
# IsFull: Check if the queue is full
# Peek: Get the value of the front of the queue without removing

class Queue:
    def __init(self):
        self.queue = []
        

File: queue.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class Queue {
    constructor() {
        this.first = null;
        this.size = 0;
    }
    enqueue(val) {
        let newNode = new Node(val)
        if(!this.first) {
            this.first = newNode;
            this.last = newNode;
        } else {
            this.last.next = newNode;
            this.last = newNode
        }
        return ++this.size;
    }

    dequeue() {
        if(!this.first) return null

        let temp = this.first
        if (this.first === this.last) {
            this.last = null;
        }
        this.first = this.first.next
        this.size--
        return temp.value
    }
}

const quickQueue = new Queue
quickQueue.enqueue(1)
quickQueue.enqueue(2)
quickQueue.enqueue(3)
console.log(quickQueue.first)
console.log(quickQueue.last)
console.log(quickQueue.size)
quickQueue.enqueue(4)
console.log(quickQueue.dequeue())